# **Extern: TruBridge Data Analytics Project**

**Question:** Which crime and school behavioral exposures are associated with the highest levels of fear among students?

## **Step 0: Data Loading:**

### **Loading Packages**

In [ ]:
##Packages:
import pandas as pd
import numpy as np

!pip install pyreadstat
import pyreadstat

##Importing files:
#from google.colab import files
#uploaded = files.upload()

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.6/2.6 MB 15.6 MB/s eta 0:00:00


In [ ]:
##Converting .sav to df:
df, meta = pyreadstat.read_sav("/content/38666-0001-Data.sav")

### **Extraction/Renaming Columns**

In [ ]:
##Extracting ONLY relevant columns:
relevant_col = [
    # Demographics
    "V3014", "V3018",
    "V3085", "V3023A", "V3024A",
    "V3020", "V3084",

    # School Information
    "VS0017", "SCS215", "SCS216", "SCS217",

    # School Environment & Safety
    "SCS189", "SCS213",

    # Bullying
    "VS0073", "VS0074", "VS0075", "VS0076", "VS0077",
    "VS0078", "VS0079", "VS0080",
    "VS0088",

    # Gangs
    "VS0133", "VS0134", "VS0135",

    # Drugs & Alcohol
    "VS0058", "VS0059", "SCS293", "VS0067",
    "SCS209", "SCS210", "SCS292",

    # Weapons
    "VS0128", "VS0130", "VS0131", "VS0132",

    # Hate Incidents
    "VS0105","VS0112",

    # Target Variable
    "VS0124",

    # Attendance & Academics
    "VS0137", "VS0138",

    # Survey Weights
    "VS0142"
]

In [ ]:
df = df[relevant_col]

In [ ]:
##Renaming Columns:
col_names = {
    # Demographics
    "V3014": 'age',
    "V3018": 'sex',
    "V3085": 'birth_gender',
    "V3023A": 'race',
    "V3024A": 'hispanic_origin',
    "V3020": 'education',
    "V3084": 'sex_orientation',

    # School Information
    "VS0017": 'current_grade',
    "SCS215": 'school_type',
    "SCS216": 'school_locale',
    "SCS217": 'school_level',

    # School Environment & Safety
    "SCS189": 'safe_at_school',
    "SCS213": 'crime_in_neighborhood',

    # Bullying
    "VS0073": 'name_calling',
    "VS0074": 'rumors',
    "VS0075": 'threats',
    "VS0076": 'physical_violence',
    "VS0077": 'forced_actions',
    "VS0078": 'social_exclusion',
    "VS0079": 'property_crime',
    "VS0080": 'days_bullied',
    "VS0088": 'reported_bullying',

    # Gangs
    "VS0133": 'school_gang',
    "VS0134": 'gang_violence',
    "VS0135": 'gang_drug_sales',

    # Drugs & Alcohol
    "VS0058": 'alcohol_available',
    "VS0059": 'marijuana_available',
    "SCS293": 'painkillers_available',
    "VS0067": 'illegal_drugs',
    "SCS209": 'other_drugs',
    "SCS210": 'saw_students_using_drugs',
    "SCS292": 'details_on_drugs',

    # Weapons
    "VS0128": 'knife_at_school',
    "VS0130": 'students_with_guns',
    "VS0131": 'saw_gun_at_school',
    "VS0132": 'access_to_gun',

    # Hate Incidents
    "VS0105": 'called_hate_speech',
    "VS0112": 'graffiti',

    # Target Variable
    "VS0124": 'fear_at_school',

    # Attendance & Academics
    "VS0137": 'days_skipped',
    "VS0138": 'grades',

    # Survey Weights
    "VS0142": 'scs_weight'
}

df = df.rename(columns=col_names)

In [ ]:
## Converting col. names:
#for var, label in zip(meta.column_names, meta.column_labels):
#    df.rename(columns={var: label}, inplace=True)

In [ ]:
##Peaking at the dataset:
df.head()

,age,sex,birth_gender,race,hispanic_origin,education,sex_orientation,current_grade,school_type,school_locale,...,knife_at_school,students_with_guns,saw_gun_at_school,access_to_gun,called_hate_speech,graffiti,fear_at_school,days_skipped,grades,scs_weight
0,17.0,1.0,NaN,1.0,2.0,11.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.000000
1,15.0,2.0,NaN,1.0,2.0,9.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.000000
2,12.0,2.0,NaN,1.0,2.0,6.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.000000
3,18.0,2.0,NaN,6.0,2.0,11.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.000000
4,13.0,1.0,NaN,6.0,2.0,6.0,NaN,2.0,8.0,8.0,...,2.0,1.0,2.0,2.0,1.0,1.0,1.0,NaN,2.0,8563.266305


In [ ]:
##Summary:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12335 entries, 0 to 12334
Data columns (total 42 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   age                       12335 non-null  float64
 1   sex                       12335 non-null  float64
 2   birth_gender              2367 non-null   float64
 3   race                      12335 non-null  float64
 4   hispanic_origin           12335 non-null  float64
 5   education                 12335 non-null  float64
 6   sex_orientation           2367 non-null   float64
 7   current_grade             4920 non-null   float64
 8   school_type               4837 non-null   float64
 9   school_locale             4837 non-null   float64
 10  school_level              4837 non-null   float64
 11  safe_at_school            4837 non-null   float64
 12  crime_in_neighborhood     4837 non-null   float64
 13  name_calling              4837 non-null   float64
 14  rumors

For simplicity, we will also quickly map all of the float values to their actual categorical values now:

## **Step 1: Data Cleaning**

### **Handling Null Values**

In [ ]:
##Checking Null Values:
print(df.isnull().sum())

age                             0
sex                             0
birth_gender                 9968
race                            0
hispanic_origin                 0
education                       0
sex_orientation              9968
current_grade                7415
school_type                  7498
school_locale                7498
school_level                 7498
safe_at_school               7498
crime_in_neighborhood        7498
name_calling                 7498
rumors                       7498
threats                      7498
physical_violence            7498
forced_actions               7498
social_exclusion             7498
property_crime               7498
days_bullied                11403
reported_bullying           11403
school_gang                  7634
gang_violence               12050
gang_drug_sales             12050
alcohol_available            7634
marijuana_available          7634
painkillers_available        7634
illegal_drugs                7634
other_drugs   

In [ ]:
##Dropping entries with missing target value:

#Copying dataset, one with no null vals:
df = df[df['fear_at_school'].notna().copy()]

df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4701 entries, 4 to 12333
Data columns (total 42 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   age                       4701 non-null   float64
 1   sex                       4701 non-null   float64
 2   birth_gender              1602 non-null   float64
 3   race                      4701 non-null   float64
 4   hispanic_origin           4701 non-null   float64
 5   education                 4701 non-null   float64
 6   sex_orientation           1602 non-null   float64
 7   current_grade             4596 non-null   float64
 8   school_type               4701 non-null   float64
 9   school_locale             4701 non-null   float64
 10  school_level              4701 non-null   float64
 11  safe_at_school            4701 non-null   float64
 12  crime_in_neighborhood     4701 non-null   float64
 13  name_calling              4701 non-null   float64
 14  rumors      

In [ ]:
#Checking which cols. should be dropped AFTER ensuring target variable is not null:
df.isnull().sum().sort_values()

,0
age,0
sex,0
race,0
hispanic_origin,0
education,0
school_level,0
school_locale,0
school_type,0
crime_in_neighborhood,0
name_calling,0


We have a total of 4701 entries. With that being said, we will drop the following columns given the large amount of null values:
- details_on_drugs
- saw_gun_at_school
- gang_drug_sales
- gang_violence
- days_skipped
- days_bullied
- birth_gender
- sex_orientation
- reported_bullying

In [ ]:
drop_col = {
"details_on_drugs",
"saw_gun_at_school",
"gang_drug_sales",
"gang_violence",
"days_skipped",
"days_bullied",
"birth_gender",
"sex_orientation",
"reported_bullying"
}

df = df.drop(columns=drop_col)

In [ ]:
## Dropping remaining NULL values in current_grade:
df = df[df['current_grade'].notna()]

In [ ]:
df.isnull().sum().sort_values()

,0
age,0
sex,0
race,0
hispanic_origin,0
education,0
current_grade,0
school_type,0
school_locale,0
school_level,0
safe_at_school,0


While we deleted a whole category (relating to gangs/gang violence), our conclusions for this would've likely been inaccurate. Therefore, instead of making a likely "inaccurate" conclusion, we decided to delete it all together.

In [ ]:
## Dropping Duplicates:
duplicates = df.duplicated()
print(duplicates.sum())

df = df.drop_duplicates()

6


### **Remapping Values**

We need to be remap columns based on the codebook's values (essentially, turning the integers into strings).

In [ ]:
mappings = {
    "sex": {1: "Male", 2: "Female", 8: "Residue", 9: "Missing"},

    "race": {1: "White only", 2: "Black only", 3: "American Indian/Alaskan Native only", 4: "Asian only", 5: "Hawaiian/Pacific Islander only", 6: "White-Black", 7: "White-American Indian", 8: "White-Asian", -1: "Not Applicable", 99: "Missing / Not Asked"},

    "hispanic_origin": {1: "Yes - Hispanic", 2: "No - Not Hispanic", 8: "Residue", -1: "Not Applicable", 9: "Missing / Not Asked"},

    "education": {0: "Never attended/Kindergarten only", 1: "1st Grade", 2: "2nd Grade", 3: "3rd Grade", 4: "4th Grade", 5: "5th Grade", 6: "6th Grade", 7: "7th Grade", 8: "8th Grade", 9: "9th Grade", 10: "10th Grade", 11: "11th Grade", 12: "12th Grade/High School Graduate", 13: "Some College/Associate's Degree", 14: "Bachelor's Degree", 15: "Graduate/Professional Degree", 99: "Missing / Not Asked"},

    "current_grade": {0: "Fifth grade or under", 1: "Sixth grade", 2: "Seventh grade", 3: "Eighth grade", 4: "Ninth grade", 5: "Tenth grade", 6: "Eleventh grade", 7: "Twelfth grade", 8: "Other grade", 9: "College/GED/Post-graduate/Non-eligible", 98: "Residue", 98.0: "Residue", 999: "Don't Know", 999.0: "Don't Know", 99: "Out of Universe / Missing", 99.0: "Out of Universe / Missing"},

    "school_type": {1: "Public school", 2: "Private school, no religious affiliation", 3: "Private, Roman Catholic school", 4: "Private, other religious school", 5: "Private, nonsectarian school", 8: "Residue", 9: "Missing / Out of Universe"},

    "school_locale": {1: "City", 2: "Suburb", 3: "Town", 4: "Rural", 5: "Missing", 8: "Residue", 9: "Missing / Out of Universe"},

    "school_level": {1: "Elementary school", 2: "Middle school", 3: "Secondary school", 4: "High school", 5: "Other", 6: "Missing", 7: "Not Applicable", 8: "Residue", 9: "Missing / Out of Universe"},

    "safe_at_school": {1: "Strongly Agree", 2: "Agree", 3: "Disagree", 4: "Strongly Disagree", 8: "Residue", 98: "Refused", 99: "Don't Know", 9: "Missing / Out of Universe"},

    "crime_in_neighborhood": {1: "Strongly Agree", 2: "Agree", 3: "Disagree", 4: "Strongly Disagree", 8: "Residue", 98: "Refused", 99: "Don't Know", 9: "Missing / Out of Universe"},

    "name_calling": {1: "Yes", 2: "No", 8: "Residue", 98: "Refused", 99: "Don't Know", 9: "Missing / Out of Universe"},

    "rumors": {1: "Yes", 2: "No", 8: "Residue", 98: "Refused", 99: "Don't Know", 9: "Missing / Out of Universe"},

    "threats": {1: "Yes", 2: "No", 8: "Residue", 98: "Refused", 99: "Don't Know", 9: "Missing / Out of Universe"},

    "physical_violence": {1: "Yes", 2: "No", 8: "Residue", 98: "Refused", 99: "Don't Know", 9: "Missing / Out of Universe"},

    "forced_actions": {1: "Yes", 2: "No", 8: "Residue", 98: "Refused", 99: "Don't Know", 9: "Missing / Out of Universe"},

    "social_exclusion": {1: "Yes", 2: "No", 8: "Residue", 98: "Refused", 99: "Don't Know", 9: "Missing / Out of Universe"},

    "property_crime": {1: "Yes", 2: "No", 8: "Residue", 98: "Refused", 99: "Don't Know", 9: "Missing / Out of Universe"},

    "school_gang": {1: "Yes", 2: "No", 3: "Don't Know", 8: "Residue", 98: "Refused", 9: "Missing / Out of Universe"},

    "alcohol_available": {1: "Yes", 2: "No", 8: "Residue", 98: "Refused", 99: "Don't Know", 9: "Missing / Out of Universe"},

    "marijuana_available": {1: "Yes", 2: "No", 8: "Residue", 98: "Refused", 99: "Don't Know", 9: "Missing / Out of Universe"},

    "painkillers_available": {1: "Yes", 2: "No", 8: "Residue", 98: "Refused", 99: "Don't Know", 9: "Missing / Out of Universe"},

    "illegal_drugs": {1: "Yes", 2: "No", 8: "Residue", 98: "Refused", 99: "Don't Know", 9: "Missing / Out of Universe"},

    "other_drugs": {1: "Yes", 2: "No", 8: "Residue", 98: "Refused", 99: "Don't Know", 9: "Missing / Out of Universe"},

    "saw_students_using_drugs": {1: "Yes", 2: "No", 8: "Residue", 98: "Refused", 99: "Don't Know", 9: "Missing / Out of Universe"},

    "knife_at_school": {1: "Yes", 2: "No", 8: "Residue", 98: "Refused", 99: "Don't Know", 9: "Missing / Out of Universe"},

    "students_with_guns": {1: "Yes", 2: "No", 8: "Residue", 98: "Refused", 99: "Don't Know", 9: "Missing / Out of Universe"},

    "access_to_gun": {1: "Yes", 2: "No", 8: "Residue", 98: "Refused", 99: "Don't Know", 9: "Missing / Out of Universe"},

    "called_hate_speech": {1: "Yes", 2: "No", 8: "Residue", 98: "Refused", 99: "Don't Know", 9: "Missing / Out of Universe"},

    "graffiti": {1: "Yes", 2: "No", 8: "Residue", 98: "Refused", 99: "Don't Know", 9: "Missing / Out of Universe"},

    "fear_at_school": {1: "Never", 2: "Almost never", 3: "Sometimes", 4: "Most of the time", 8: "Residue", 98: "Refused", 99: "Don't Know", 9: "Missing / Out of Universe"},

    "grades": {1: "A's", 2: "B's", 3: "C's", 4: "D's", 5: "F's", 6: "passes", 7: "fails", 8: "School does not give grades / no alphabetic equivalent", 98: "Residue", 998: "Refused", 999: "Don't Know", 99: "Missing / Out of Universe"},

    "scs_weight": None
}

for col_name, mapping in mappings.items():
    if col_name in df.columns and mapping is not None:
        df[col_name] = df[col_name].map(mapping).astype('string')

## Numpy:

### **Deleting Invalid Values**

In [ ]:
missing_labels = [
    "Residue",
    "Missing",
    "Missing / Not Asked",
    "Missing / Out of Universe",
    "Don't Know",
    "Refused"
]
df = df.replace(missing_labels, np.nan)

In [ ]:
df.isnull().sum().sort_values()

,0
age,0
sex,0
hispanic_origin,0
scs_weight,0
current_grade,16
race,37
school_level,37
safe_at_school,48
property_crime,57
physical_violence,59


In [ ]:
## Given we already dropped the previous cols related to gang violence and it's high null count, we will drop the last one tied to it:
df = df.drop(columns = 'school_gang')

### **Imputing Mode to all values with NAs**

In [ ]:
##Imputting the mode for columns with missing values:
impute_cols = ['school_locale', 'school_type', 'school_type','other_drugs', 'illegal_drugs', 'painkillers_available', 'marijuana_available', 'alcohol_available',
                'saw_students_using_drugs', 'saw_students_using_drugs', 'access_to_gun', 'students_with_guns','grades', 'graffiti', 'knife_at_school', 'rumors', 'called_hate_speech', 'fear_at_school',
               'name_calling', 'social_exclusion', 'forced_actions', 'threats', 'physical_violence','property_crime', 'safe_at_school', 'school_level', 'race', 'current_grade'
               ]

for col in impute_cols:
  mode_val = df[col].mode()[0]
  df[col].fillna(mode_val, inplace=True)

/tmp/ipykernel_1047/443095307.py:9: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna(mode_val, inplace=True)


### **Remapping as Integers**

In [ ]:
## Turning "Yes/no" columns back into binary 1/0
# List of all the columns you want to convert
binary_cols = [
    "name_calling", "rumors", "threats", "physical_violence",
    "forced_actions", "social_exclusion", "property_crime",
    "alcohol_available", "marijuana_available",
    "painkillers_available", "illegal_drugs", "other_drugs",
    "saw_students_using_drugs", "knife_at_school", "students_with_guns",
    "access_to_gun", "called_hate_speech", "graffiti"
]

binary_map = {'Yes': 1, 'No': 0}

for col in binary_cols:
    df[col] = df[col].map(binary_map)

In [ ]:
## We will do the same for "scaled" questions (Strongly Negative -> Strong Positive)

#Flipping the scale used by the codebook:
scale_map = {"Strongly Disagree": 1, "Disagree": 2,"Agree": 3,"Strongly Agree": 4}
scale_cols = ["crime_in_neighborhood", "safe_at_school"]

for col in scale_cols:
    df[col] = df[col].map(scale_map)

In [ ]:
## Same logic will be applied to fear_at_school:
fear_scale = {"Never": 1, "Almost never": 2,"Sometimes": 3, "Most of the time": 4}

df['fear_at_school'] = df['fear_at_school'].map(fear_scale)


### **Numpy Operations**

In [ ]:
## Checking for outliers:

#Only one numeric column, so we check ages:
df['age'].value_counts()

,count
age,
13.0,748
14.0,746
15.0,703
12.0,689
17.0,683
16.0,657
18.0,364


In [ ]:
## Calculating descriptive statistics for age:
age_len = df['age'].values

avg_age = np.mean(age_len)
med_age = np.median(age_len)

print('Average age', avg_age)
print('Median age', med_age)

Average age 14.73986928104575
Median age 15.0


Seems like our data could be slightly skewed towards younger children

In [ ]:
## Normalizing Ages:
min_age = np.min(age_len)
max_age = np.max(age_len)

df['norm_age'] = (df['age'] - min_age) / (max_age - min_age)

In [ ]:
df['norm_age']

,norm_age
4,0.166667
9,0.500000
13,0.666667
21,0.000000
22,0.333333
...,...
12316,0.333333
12328,0.000000
12330,0.500000
12331,0.666667


### **Final Touch Ups**

In [ ]:
df.isnull().sum().sort_values()

,0
age,0
sex,0
race,0
hispanic_origin,0
current_grade,0
school_type,0
school_locale,0
school_level,0
physical_violence,0
safe_at_school,0


In [ ]:
##Imputting mode after transformations:
impute_2cols = {"crime_in_neighborhood", "education"}

for col in impute_2cols:
  mode_val = df[col].mode()[0]
  df[col].fillna(mode_val, inplace=True)

/tmp/ipykernel_1047/3341856405.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna(mode_val, inplace=True)


In [ ]:
## Changing "crime_in_neighborhood" & "age" as int
df['crime_in_neighborhood'] = df['crime_in_neighborhood'].astype('Int64')
df['age'] = df['age'].astype('Int64')

In [ ]:
##Dropping duplicates in case there are any remaining:
df = df.drop_duplicates()

### **Final Dataset**

In [ ]:
df.head()

,age,sex,race,hispanic_origin,education,current_grade,school_type,school_locale,school_level,safe_at_school,...,saw_students_using_drugs,knife_at_school,students_with_guns,access_to_gun,called_hate_speech,graffiti,fear_at_school,grades,scs_weight,norm_age
4,13,Male,White-Black,No - Not Hispanic,6th Grade,Seventh grade,Public school,Suburb,Middle school,3,...,0,0,1,0,1,1,1,B's,8563.266305,0.166667
9,15,Male,White only,Yes - Hispanic,10th Grade,Tenth grade,Public school,Suburb,High school,3,...,0,0,0,0,0,0,1,B's,9392.339515,0.500000
13,16,Male,White only,No - Not Hispanic,8th Grade,Tenth grade,Public school,City,High school,4,...,0,0,0,0,0,0,1,A's,4251.380709,0.666667
21,12,Male,White only,Yes - Hispanic,6th Grade,Sixth grade,Public school,Suburb,Elementary school,3,...,0,0,0,0,0,0,2,B's,7012.472992,0.000000
22,14,Male,White only,Yes - Hispanic,8th Grade,Eighth grade,Public school,Suburb,Middle school,4,...,1,0,0,0,0,1,1,B's,5549.336008,0.333333


In [ ]:
df.columns

Index(['age', 'sex', 'race', 'hispanic_origin', 'education', 'current_grade',
       'school_type', 'school_locale', 'school_level', 'safe_at_school',
       'crime_in_neighborhood', 'name_calling', 'rumors', 'threats',
       'physical_violence', 'forced_actions', 'social_exclusion',
       'property_crime', 'alcohol_available', 'marijuana_available',
       'painkillers_available', 'illegal_drugs', 'other_drugs',
       'saw_students_using_drugs', 'knife_at_school', 'students_with_guns',
       'access_to_gun', 'called_hate_speech', 'graffiti', 'fear_at_school',
       'grades', 'scs_weight', 'norm_age'],
      dtype='object')

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4589 entries, 4 to 12333
Data columns (total 33 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   age                       4589 non-null   Int64  
 1   sex                       4589 non-null   string 
 2   race                      4589 non-null   string 
 3   hispanic_origin           4589 non-null   string 
 4   education                 4589 non-null   string 
 5   current_grade             4589 non-null   string 
 6   school_type               4589 non-null   string 
 7   school_locale             4589 non-null   string 
 8   school_level              4589 non-null   string 
 9   safe_at_school            4589 non-null   int64  
 10  crime_in_neighborhood     4589 non-null   Int64  
 11  name_calling              4589 non-null   int64  
 12  rumors                    4589 non-null   int64  
 13  threats                   4589 non-null   int64  
 14  physical_vio

In [ ]:
## Exporting:
# np.savetxt('cleaned_school.csv', np.column_stack(('age', 'sex', 'race', 'hispanic_origin', 'education', 'current_grade',
#        'school_type', 'school_locale', 'school_level', 'safe_at_school',
#        'crime_in_neighborhood', 'name_calling', 'rumors', 'threats',
#        'physical_violence', 'forced_actions', 'social_exclusion',
#        'property_crime', 'alcohol_available', 'marijuana_available',
#        'painkillers_available', 'illegal_drugs', 'other_drugs',
#        'saw_students_using_drugs', 'knife_at_school', 'students_with_guns',
#        'access_to_gun', 'called_hate_speech', 'graffiti', 'fear_at_school',
#        'grades', 'scs_weight', 'norm_age')), delimiter=",", fmt='%s', header="School Crime csv")


#For simplicity, we'll use df.to_csv:
df.to_csv('cleaned_school.csv', index = False)